service/policy/adr/0002-pagination-list-rpcs.md

```diff
@@ -0,0 +1,175 @@
+}
+```
+
+```sql
+-- subject-mappings example request:
+--   'after' 2023-01-01
+--   'limit' 100
+SELECT * FROM opentdf_policy.subject_mappings
+WHERE created_at > '2023-01-01' ORDER BY created_at LIMIT 100;
+```
+
+#### Pros & Cons
+
+- :green_circle: Support - supported across any SQL database (just slightly different syntax)
+- :green_circle: Speed - much faster in deep pages than OFFSET due to reduced scan row count
+- :yellow_circle: Reliability - provisioned policy may contain the same `created_at` timestamp
+- :red_circle: Flexibility - pagination is only forward of the `created_at` timestamp
+- :red_circle: Complexity - client must maintain state since response timestamps are required to drive subsequent request timestamp pagination, and pagination backwards is not supported
+- :red_circle: Complexity - reliance on timestamps introduces timezone differential confusion unless a parameter is also employed to localize the query
+
+### Cursor Pagination
+
+We would index a column (the most obvious would be `created_at`) to use as the pagination key for
+querying, but we would utilize an encoded cursor approach.
+
+```proto
+message ListRequest {
+    // ...existing fields omitted
+    int32 limit = 3; // default depends on type of policy object
+    string cursor = 4; // defaulted in API layer to cursor for encoded start_of_time
+}
+
+message ListResponse {
+    // ...existing fields and response data ommitted
+    // cursors are encoded by the server as base64'd 'created_at' timestamps
+    string previous_cursor = 4;
+    string next_cursor = 4;
+    int32 total = 5; // indication of total available for pagination
+}
+```
+
+```sql
+-- subject-mappings example, request:
+--   'after_cursor' 2023-01-01 00:00:00.000000+00
+--   'limit' 100
+WITH Data AS (
+    SELECT *
+    FROM opentdf_policy.subject_mappings
+    WHERE created_at >= '2023-01-01 00:00:00.000000+00'
+    ORDER BY created_at
+    LIMIT 101
+),
+NextPage AS (
+    SELECT *
+    FROM Data
+    ORDER BY created_at
+    LIMIT 100
+),
+PreviousPage AS (
+    SELECT *
+    FROM opentdf_policy.subject_mappings
+    WHERE created_at < (SELECT MIN(created_at) FROM Data)
+    ORDER BY created_at DESC
+    LIMIT 101
+),
+CursorData AS (
+    SELECT
+        (SELECT MIN(created_at) FROM Data) AS first_item_created_at,
+        (SELECT MAX(created_at) FROM NextPage) AS next_cursor_created_at,
+        (SELECT MIN(created_at) FROM PreviousPage) AS previous_cursor_created_at
+)
+SELECT
+    (SELECT json_agg(row_to_json(NextPage)) FROM NextPage) AS data,
+    (SELECT json_build_object('created_at', next_cursor_created_at) FROM CursorData) AS next_cursor,
+    (SELECT json_build_object('created_at', previous_cursor_created_at) FROM CursorData) AS previous_cursor
+FROM CursorData;
+```
+
+#### Pros & Cons
+
+- :green_circle: Support - supported across any SQL database (just different syntax)
+- :green_circle: Speed - much faster in deep pages than OFFSET due to reduced scan row count
```

Comment:
Negative is that we have to index the date which probably has little practical value with day to day use or management. 

Adding attributes is not like adding users since they are grouped by namespace and definitions.